teps/0057-windows-support.md

```diff
@@ -0,0 +1,154 @@
+
+## Motivation
+Our specific use case is building Unreal Engine projects in containers. It is not possible to build UE projects for Windows on Linux, so we must be able to build in Windows containers, which in turn requires support for Windows nodes. There are other scenarios which require the use of Windows nodes/containers and this TEP is to support any Windows-based use case. There is currently no viable alternative approach available here (see alternatives below). If we have a TaskRun that requires a Windows container then it must run on a Windows node. Currently this is an unsupported use case, and the goal of this TEP is to address this need.
+
+### Goals
+- Allow Tekton TaskRuns to selectively run on Windows nodes
+- Allow a TaskRun to use nodeSelector to request this feature
+- Core Tekton components (Operator, CLI and Dashboard) must be able to recognise and communicate with Windows nodes in the same way they do Linux nodes
+- Tasks on the Catalog and Hub should be able to indicate the requirement for a Windows node to run it
+- All current (Linux-only) workflows should remain unchanged
+
+### Non-Goals
+- The ability to run core Tekton components on Windows nodes is not a goal of this TEP
+    - It is expected that users will have a mix of Linux and Windows nodes and that the core Tekton controller components will run on Linux nodes
+- Tekton will not be responsible for determining whether a Pod for a TaskRun should be created on a Windows node. This will be an opt-in feature determined via a nodeSelector.
+
+
+<!-- ### Use Cases (optional) -->
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+## Requirements
+Tekton should be able to:
+- Recognise and communicate with Windows nodes in the cluster
+- Allow TaskRuns to specify the need to run on a Windows node
+- Create Pods for TaskRuns on Windows nodes
+- Provide full support for all Tekton features to Windows based TaskRuns
+- Manage Pipelines containing a mix of Windows and Linux based Tasks
+
+Workflows that do not require Windows support should not be changed in any way.
+
+Tekton Hub and Catalog should allow a Task to indicate that it requires a Windows node to run, if this is absent, the default behaviour will be to assume Linux so that no change to existing Linux workflows is required.
+
+## Proposal
+Extend Tekton so TaskRuns can request, and be scheduled, to run on Windows nodes. This will require (at least) reworking the following Tekton components:
+- Entrypoint binary
+- All internal support images used for resource management
+- File paths used for step ordering, either via polling or watching files
+- Script mode
```

Comment:
@sbwsg pointed out in the [API working group notes](https://docs.google.com/document/d/17PodAxG8hV351fBhSu7Y_OIPhGTVgj6OJ2lPphYYRpU/edit#) that [pkg/pod/pod.go](https://github.com/tektoncd/pipeline/blob/main/pkg/pod/pod.go) might be a hotspot for windows related changes